File: leetcode/python/P0063_UniquePaths2.py
# https://leetcode.com/problems/unique-paths-ii/

from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        # keep a grid of number of ways
        # no need to pad; set 1 on all items on first row and column
        # at index (i, j) inside the dist matrix:
        #   ways[i, j] = ways[i, j-1] + ways[i - 1][j]
        # but it becomes 0 if thee is an obstacle.
        # Pay attention to obstacle at (0, 0).

        n_rows = len(obstacleGrid)
        # there is at least one element
        n_cols = len(obstacleGrid[0])

        ways = [[0 for _ in range(n_cols)] for _ in range(n_rows)]

        for r in range(n_rows):
            for c in range(n_cols):
                ways[r][c] = 0
                if r == 0 and c == 0:
                    ways[r][c] = 1
                else:
                    if c > 0:
                        ways[r][c] += ways[r][c - 1]
                    if r > 0:
                        ways[r][c] += ways[r - 1][c]
                ways[r][c] *= 1 - obstacleGrid[r][c]

        return ways[-1][-1]

File: leetcode/cpp/P0063_UniquePaths2.cpp
// https://leetcode.com/problems/unique-paths-ii/
// https://leetcode.com/problems/unique-paths-ii

#include <array>
#include <iostream>
#include <vector>

class Solution {
 public:
  int uniquePathsWithObstacles(
      const std::vector<std::vector<int>>& obstacleGrid) {
    int n_rows = obstacleGrid.size();
    int n_cols = obstacleGrid[0].size();

    // can preallocate, because grid is small
    // and we don't need the overhead from vector.
    std::array<std::array<int, 128>, 128> ways;
    for (int row = 0; row < n_rows; ++row) {
      ways[row].fill(0);
    }

    for (int row = 0; row < n_rows; ++row) {
      for (int col = 0; col < n_cols; ++col) {
        if (row == 0 && col == 0) {
          ways[row][col] = 1;
        } else {
          if (col > 0) {
            ways[row][col] += ways[row][col - 1];
          }
          if (row > 0) {
            ways[row][col] += ways[row - 1][col];
          }
        }
        ways[row][col] *= (1 - obstacleGrid[row][col]);
      }
    }

    return ways[n_rows - 1][n_cols - 1];
  }
};

int main(int argc, char** argv) {
  Solution s;
  std::cout << s.uniquePathsWithObstacles({{0, 0, 0}, {0, 1, 0}, {0, 0, 0}})
            << std::endl;
  return 0;
}